src/main/scala/org/broadinstitute/dsde/workbench/leonardo/Boot.scala

```diff
@@ -93,6 +92,14 @@ object Boot extends App with LazyLogging {
       .asInstanceOf[T]
   }
 
+
+  private def constructProvider[T](className: String,  config: Config, serviceAccountProvider: ServiceAccountProvider): T = {
+    Class.forName(className)
+      .getConstructor(classOf[Config], classOf[ServiceAccountProvider])
```

Comment:
This assumes a constructor that takes a `Config` and a `ServiceAccountProvider`. This seems kind of specialized to `SamAuthProvider` -- seems like it would fail to create a `PetServiceAccountProvider`?

I am wondering if this would work as a generic way to instantiate a `Provider` object via reflection:
```
  private def constructProvider[T](className: String, ctorArgs: Any*): T = {
    Class.forName(className)
      .getConstructor(ctorArgs.map(_.getClass.getInterfaces()(0)):_*)
      .newInstance(ctorArgs)
      .asInstanceOf[T]
  }
```

Otherwise you could have 2 methods: `constructServiceAccountProvider` (which takes a `Config`) and `constructLeoAuthProvider` (which takes a `Config` and a `ServiceAccountProvider`).

Reply:
I assumed your last paragraph -- that we'd just break up `constructProvider[T]` into separate SA-provider and auth-provider ones.

Reply:
Yeah, I was thinking that calling `.getInterfaces()` might fix the generic version... but I am ok with un-generifying `constructProvider`.

Reply:
Oh sorry, I missed that this _is_ a separate `constructProvider` implementation. Since we have separate methods we should name them separately and remove the generic type param. So:
```
private def constructServiceAccountProvider(className: String, config: Config): ServiceAccountProvider = {
   Class.forName(className)
     .getConstructor(classOf[Config])
     .newInstance(config)
      .asInstanceOf[ServiceAccountProvider]
  }

  private def constructLeoAuthProvider(className: String,  config: Config, serviceAccountProvider: ServiceAccountProvider): LeoAuthProvider = {
    Class.forName(className)
      .getConstructor(classOf[Config], classOf[ServiceAccountProvider])
      .newInstance(config, serviceAccountProvider)
      .asInstanceOf[T]
  }
```